onnxruntime/core/providers/cpu/math/einsum_utils.cc

```diff
@@ -0,0 +1,724 @@
+template <typename T>
+Status EinsumTypedComputeProcessor(OpKernelContext* context,
+                                   AllocatorPtr allocator,
+                                   EinsumComputePreprocessor& einsum_compute_preprocessor) {
+  const auto& mapped_indices_to_last_input_index = einsum_compute_preprocessor.GetMappedSubscriptIndicesToLastInputIndex();
+
+  auto& preprocessed_inputs = einsum_compute_preprocessor.GetPreprocessedTensors();
+
+  auto num_subscript_labels = einsum_compute_preprocessor.GetNumSubscriptIndices();
+
+  const auto& output_dims = einsum_compute_preprocessor.GetOutputDims();
+
+  auto* output = context->Output(0, output_dims);
+
+  auto num_inputs = context->InputCount();
+
+  concurrency::ThreadPool* tp = context->GetOperatorThreadPool();
+
+  // Preprocess the first input so as to reduce any dims that only it has
+  Tensor result;
+
+  {
+    std::vector<int64_t> reduced_dims;
+    std::vector<int64_t> preserved_dims;            // dims which were not reduced
+    std::vector<int64_t> preserved_shape;           // shape pertaining to only the dims that were preserved (not reduced)
+    reduced_dims.reserve(num_subscript_labels);     // num_subscript_labels is the upper bound. No harm in over-reserving.
+    preserved_dims.reserve(num_subscript_labels);   // num_subscript_labels is the upper bound. No harm in over-reserving.
+    preserved_shape.reserve(num_subscript_labels);  // num_subscript_labels is the upper bound. No harm in over-reserving.
+
+    auto& first_input = preprocessed_inputs[0];
+    const auto& dims = first_input.Shape().GetDims();
+    for (int64_t i = 0; i < num_subscript_labels; ++i) {
+      if (mapped_indices_to_last_input_index[i] == 0) {
+        reduced_dims.push_back(i);
+      } else {
+        preserved_dims.push_back(i);
+        preserved_shape.push_back(dims[i]);
+      }
+    }
+
+    // Reduce the dims that are last seen in the first input alone
+    if (reduced_dims.size() != 0) {
+      first_input = EinsumOp::ReduceSum<T>(first_input, reduced_dims, allocator, tp);
+    }
+
+    // Finalize the output at this stage if num_inputs == 1
+    if (num_inputs == 1) {
+      // Create reshaped view to squeeze out the reduced dims
+      EinsumOp::CreateReshapedView(first_input, preserved_shape);
+
+      // Finalize the output by applying any transpose required to get it to the required output ordering and move it to the op's output
+      EinsumOp::FinalizeOutput<T>(first_input, preserved_dims, einsum_compute_preprocessor.GetMappedSubscriptIndicesToOutputindices(), *output, output_dims, allocator);
+
+      return Status::OK();
+    } else {  // Assign the first tensor to result to proceed further
+      result = std::move(first_input);
+    }
+  }
+
+  // Process the operands in a pair-wise fashion
+  {
+    bool is_final_pair = false;
+    // Keep processing each input pair-wise
+    for (int input = 1; input < num_inputs; ++input) {
+      std::vector<int64_t> reduced_dims;
```

Comment:
Didn't see any unit tests with more than 2 inputs. Do we need to support that? If so, please add some tests.

Reply:
Yes. Added some tests. Thanks for catching.